core/kotlinx-coroutines-core/src/main/kotlin/kotlinx/coroutines/experimental/Await.kt

```diff
@@ -0,0 +1,66 @@
+package kotlinx.coroutines.experimental
+
+import kotlinx.atomicfu.atomic
+
+/**
+ * Awaits for completion of given jobs without blocking a thread and resumes normally when all deferred computations are complete
+ * or resumes with the first thrown exception if any of computations completes exceptionally including cancellation.
+ * This suspending function is cancellable.
+ * If the [Job] of the current coroutine is cancelled or completed while this suspending function is waiting,
+ * this function immediately resumes with [CancellationException].
+ */
+public suspend fun awaitAll(vararg jobs: Job): Unit = jobs.asList().awaitAll()
+
+/**
+ * Awaits for completion of given jobs without blocking a thread and resumes normally when all jobs computations are complete
+ * or resumes with the first thrown exception if any of computations completes exceptionally including cancellation.
+ * This suspending function is cancellable.
+ * If the [Job] of the current coroutine is l or completed while this suspending function is waiting,
+ * this function immediately resumes with [CancellationException].
+ */
+public suspend fun Iterable<Job>.awaitAll(): Unit {
+    val jobs = this as? Collection<Job> ?: this.toList()
```

Comment:
What if `jobs` collection is changed after calling `.awaitAll()` extension on it?

Reply:
Depends on mutation. In the worst case, `awaitAll` will hang (if concurrent mutator removes an element while `awaitAll` prepares its handles), but a non-trivial effort is required to write such code.

Behaviour in such case is undefined, what does user expect if he creates a collection of jobs, shares it and mutates in one coroutine while awaiting them in another?

Reply:
So, semantically we want `awaitAll` to operate on a some kind of snapshot. The challenge is how far we are ready to go for an optimal implementation. The easy, but not efficient, way it so to always take a snapshot of collection first, then await on it.